o/snapstate: support user-services when refreshing snaps

Meulengracht opened the pull request:
Implements the remaining functionality to support user-services when refreshing snaps. SnapState now also keeps track of the last disabled services for user-services.

Meulengracht commented:
I dropped the spread test I was doing, we were already covering this functionality in the unit tests - and extending them to cover the user services part made sense.

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -2998,6 +3013,11 @@ func (m *SnapManager) startSnapServices(t *state.Task, _ *tomb.Tomb) error {
 
+	logger.Noticef("previously disabled system-services: %v", snapst.LastActiveDisabledServices)
+	for uid, svcs := range snapst.LastActiveDisabledUserServices {
+		logger.Noticef("previously disabled user-services [%d]: %v", uid, svcs)
```
tests never hit here

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -3013,29 +3033,39 @@ func (m *SnapManager) startSnapServices(t *state.Task, _ *tomb.Tomb) error {
+	// Insert it into the overview, into the 'found' list. Re-use the entire
+	// list for all users (-1) as there won't be any difference between users
+	// in this specific case (and to allow us to keep a singular list of services
+	// that need to be disabled).
```
this comment is not very clear, can we use a single list because hooks cannot target single users via snapctl commands?

Meulengracht replied:
Your comment let me to a missing case of support for user-services. The service tracking code for hooks did not account for user-services - I added this in here, as it is required for completeness in the refresh handling.

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -3173,6 +3215,20 @@ func (m *SnapManager) stopSnapServices(t *state.Task, _ *tomb.Tomb) error {
 		disabledServices.SystemServices...,
 	)
+	// Merge the two user-services maps, ideally we would have one struct but we must
+	// preserve the backwards-compat with the json layout
```
I'm not quite sure what is referred here by "one struct"

pedronis commented on usersession/client/client.go:
```diff
@@ -333,8 +333,19 @@ func (client *Client) ServicesDaemonReload(ctx context.Context) error {
 }
 
-func filterDisabledServices(all, disabled []string) []string {
+func filterDisabledServices(all []string, disabledSvcs map[int][]string, uid int) []string {
```
does some doc comment need updating in this package?

Meulengracht replied:
Good point! added

pedronis reviewed (commented):
some comments/questions

Meulengracht commented:
I split out the wrapper changes: https://github.com/snapcore/snapd/pull/13957

bboozzoo commented on overlord/snapstate/handlers.go:
```diff
@@ -2942,10 +2958,61 @@ func (m *SnapManager) doToggleSnapFlags(t *state.Task, _ *tomb.Tomb) error {
+	return svcsToDisable
+}
+
+func installModeDisabledUserServices(snapst *SnapState, currentInfo *snap.Info, prevCurrentSvcs map[string]bool) (map[int][]string, error) {
```
could use a comment in case it's not obvious for folks who later look at the code, that the disabled services are collected only from user sessions which are active at the time this code runs

Meulengracht replied:
Added!

bboozzoo reviewed (dismissed):
LGTM, comment tweaks mostly

Meulengracht commented:
@bboozzoo thanks for the review I'll handle the comments!

I just need some time to think over the strategy here for user-services during refresh. I think I may need to split out a couple of logic steps to provide a better implementation. I have mixed up the purpose of start/stopping and enabling / disabling concepts from a perspective of a per-user state.

Meulengracht commented on overlord/servicestate/helpers.go:
```diff
@@ -21,25 +21,70 @@ package servicestate
+	var uids []int
+	var err error
+	if len(users) == 0 {
+		uids, err = clientutil.AvailableUserSessions()
```
Good point, added better docs in https://github.com/snapcore/snapd/pull/13905/commits/0075433bdfb6cf05d3ac52652b0afa6157523854

Meulengracht commented on overlord/servicestate/helpers.go:
```diff
@@ -95,5 +139,139 @@ func updateSnapstateServices(snapst *snapstate.SnapState, enable, disable []*sna
+		return sysChanged || usrChanged, nil
+	case wrappers.ServiceScopeSystem:
+		return updateSnapstateSystemServices(snapst, sys, isEnable), nil
+	case wrappers.ServiceScopeUser:
```
Added!

Meulengracht commented:
@bboozzoo @ernestl @pedronis I've removed the blocked label, and I've added better explanations for the current approach to kind of back up the reasoning here.

Please see https://github.com/snapcore/snapd/pull/13905/commits/0075433bdfb6cf05d3ac52652b0afa6157523854 for added docs. 
When reviewing this please be *very* critical of code that **should** be better explained, as we will most likely need to come back to this code in the future once we improve our support for per-user state in snapd. I.e it's important that we can understand this again, and unfortunately this state tracking is kinda complex

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -3037,29 +3107,44 @@ func (m *SnapManager) startSnapServices(t *state.Task, _ *tomb.Tomb) error {
 
 	// check what services with "InstallMode: disable" need to be disabled
-	svcsToDisableFromInstallMode, err := installModeDisabledServices(st, snapst, currentInfo)
+	sysSvcsToDisable, usrSvcsToDisable, err := installModeDisabledServices(st, snapst, currentInfo)
```
can we add the InstallMode suffix for these?

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -3037,29 +3107,44 @@ func (m *SnapManager) startSnapServices(t *state.Task, _ *tomb.Tomb) error {
+	missingSvcsOverview.FoundSystemServices = append(missingSvcsOverview.FoundSystemServices, snapst.ServicesDisabledByHooks...)
+	// merge user services disabled by hooks
+	for uid, svcs := range snapst.UserServicesDisabledByHooks {
+		missingSvcsOverview.FoundUserServices[uid] = append(missingSvcsOverview.FoundUserServices[uid], svcs...)
```
this bit of the code is not exercised by tests

Meulengracht replied:
is now covered, thank you for spotting this

pedronis reviewed (commented):
thanks for the comments, did another pass

Meulengracht commented on client/clientutil/session_test.go:
```diff
@@ -0,0 +1,76 @@
+var _ = Suite(&sessionSuite{})
+
+func (s *sessionSuite) SetUpTest(c *C) {
+	dirs.SetRootDir(c.MkDir())
```
good point, added!

Meulengracht commented on overlord/servicestate/helpers.go:
```diff
@@ -95,5 +145,145 @@ func updateSnapstateServices(snapst *snapstate.SnapState, enable, disable []*sna
+			return false, err
+		}
+		return updateSnapstateUserServices(snapst, sys, isEnable, uids), nil
+	}
```
Doing `if` here was a much nicer solution and made things simpler, thank you for that suggestion

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -2973,10 +2989,67 @@ func (m *SnapManager) doToggleSnapFlags(t *state.Task, _ *tomb.Tomb) error {
 // seen for the first time are considered.
-func installModeDisabledServices(st *state.State, snapst *SnapState, currentInfo *snap.Info) (svcsToDisable []string, err error) {
+func installModeDisabledServices(st *state.State, snapst *SnapState, currentInfo *snap.Info) (sysSvcsToDisable []string, usrSvcsToDisable map[int][]string, err error) {
 	enabledByHookSvcs := map[string]bool{}
 	for _, svcName := range snapst.ServicesEnabledByHooks {
 		enabledByHookSvcs[svcName] = true
```
is this map used in the new code?

Meulengracht replied:
You are absolutely right - it was a leftover from moving the code!

pedronis commented on overlord/snapstate/snapmgr.go:
```diff
@@ -305,11 +305,14 @@ type SnapState struct {
 	// in the snap are removed from this list on link-snap, so that we can
 	// remember services that were disabled in another revision and then renamed
 	// or otherwise removed from the snap in a future refresh.
```
does this comment need updates?

Meulengracht replied:
I added a small separate comment for the user one, but the comment should be good

pedronis reviewed (commented):
couple of comments, also it's a bit unclear to me how is --users=all in hooks treated, as it seems this consider only users one-by-one?  but for example --enable --user=all should enable for all users?

Meulengracht commented:
> couple of comments, also it's a bit unclear to me how is --users=all in hooks treated, as it seems this consider only users one-by-one? but for example --enable --user=all should enable for all users? do we have tests for that scenario?

It's a really good question and highlights that there may be missing some documentation around the solution! If --users=all is set, then there wont be a uid filter, which means it will discover all active user agent sessions and try to affect all of them.

Meulengracht commented on overlord/servicestate/helpers.go:
```diff
@@ -21,25 +21,76 @@ package servicestate
+	return sys, usr
+}
+
+func updateSnapstateSystemServices(snapst *snapstate.SnapState, apps []*snap.AppInfo, enable bool) (changed bool) {
```
added - thank you for pointing this out

bboozzoo commented on client/clientutil/session.go:
```diff
@@ -0,0 +1,51 @@
+// AvailableUserSessions returns a list of available user-session targets for
+// snapd, by probing the available snapd-session-agent sockets in the
+// XDG runtime directory.
+func AvailableUserSessions() ([]int, error) {
```
can we call this code from usersession/client maybe?

Meulengracht replied:
good point!

bboozzoo commented on overlord/servicestate/helpers.go:
```diff
@@ -21,25 +21,80 @@ package servicestate
+
+// usernamesToUids converts a list of usernames, to a list of uids by
+// looking up the usernames.
+func usernamesToUids(usernames []string) ([]int, error) {
```
we have similar code in wrappers/services.go, would it make sense to move it to osutil?

Meulengracht replied:
moved this

bboozzoo commented on overlord/snapstate/handlers.go:
```diff
@@ -2977,22 +2993,73 @@ func (m *SnapManager) doToggleSnapFlags(t *state.Task, _ *tomb.Tomb) error {
+// with user services that have been marked for 'install-mode: disable', which
+// were not already disabled for each of the active users.
+// The reason we are doing this only for users currently logged in, is because we
+// do a best-effort handling of user sewrvices - we can only query the user service
```
```suggestion
// do a best-effort handling of user services - we can only query the user service
```

bboozzoo reviewed (dismissed):
Some suggestions

Meulengracht commented:
I rebased this PR to get the latest master to get test fixes.

pedronis commented on osutil/user.go:
```diff
@@ -404,3 +404,29 @@ func UidGid(u *user.User) (sys.UserID, sys.GroupID, error) {
+	return func() { userLookup = realUserLookup }
+}
+
+func UsernamesToUids(users []string) (map[int]string, error) {
```
this needs a doc comment

Meulengracht replied:
good point!

pedronis commented on tests/main/snap-user-services-mixed-refresh-single-disabled/task.yaml:
```diff
@@ -90,22 +90,28 @@ execute: |
+    # is by --global enable, which means that any attempt to individually disable
+    # services will be a no-op.
+    # TODO: Switch to a user-preset approach and scrap the --global enable/disable
+    # for user-services specifically
```
we probably can't scrap the global enable/disable but need to have a way to override per user, no? or just error if we get --user and --enable/--disable ?

Meulengracht replied:
I will update this comment - this is no longer really valid as we want to maybe support user masking of services instead of preset. I don't think preset is the right way to go anymore

pedronis reviewed (approved):
thanks, bit confused what is the plan forward but doesn't block the PR as is, the last commits need a re-review by @bboozzoo

bboozzoo reviewed (approved):
LGTM
